src/navigation/_CollapsingNavigationMenu.jsx

```diff
@@ -125,65 +133,142 @@ const CollapsingNavigationMenu = ({ selectedPath = undefined, menuItems, onSelec
       event.preventDefault();
       handleOnClick(event, item);
+    } else if (event.nativeEvent.keyCode === KeyCode.KEY_DOWN) {
+      event.preventDefault();
+      handleDownArrow();
+    } else if (event.nativeEvent.keyCode === KeyCode.KEY_UP) {
+      event.preventDefault();
+      handleUpArrow();
+    } else if ((event.nativeEvent.keyCode === KeyCode.KEY_RIGHT || event.nativeEvent.keyCode === KeyCode.KEY_LEFT) && !currentNodeId) {
+      event.preventDefault();
+    } else if (event.nativeEvent.keyCode === KeyCode.KEY_RIGHT) {
+      event.preventDefault();
+      if (document.getElementById(currentNodeId).ariaExpanded === 'true') {
+        handleDownArrow();
+      } else if (document.getElementById(currentNodeId).ariaHasPopup && (!document.getElementById(currentNodeId).ariaExpanded || document.getElementById(currentNodeId).ariaExpanded === 'false')) {
+        handleOnClick(event, item);
+      }
+    } else if (event.nativeEvent.keyCode === KeyCode.KEY_LEFT) {
+      event.preventDefault();
+      if (document.getElementById(currentNodeId).ariaExpanded === 'true') {
+        handleOnClick(event, item);
+      } else if (!document.getElementById(currentNodeId).ariaHasPopup || !document.getElementById(currentNodeId).ariaExpanded || document.getElementById(currentNodeId).ariaExpanded === 'false') {
+        findParent();
+      }
+    } else if (event.nativeEvent.keyCode === KeyCode.KEY_HOME) {
+      event.preventDefault();
+      cursor.current = 0;
+      setCurrentNodeId(visibleNodes.current[cursor.current].id);
+    } else if (event.nativeEvent.keyCode === KeyCode.KEY_END) {
+      event.preventDefault();
+      cursor.current = visibleNodes.current.length - 1;
+      setCurrentNodeId(visibleNodes.current[cursor.current].id);
+    } else if (event.nativeEvent.keyCode >= KeyCode.KEY_A && event.nativeEvent.keyCode <= KeyCode.KEY_Z) {
+      event.preventDefault();
+      const char = String.fromCharCode(event.nativeEvent.keyCode);
+      findNode(char);
+    } else if (event.nativeEvent.shiftKey && event.nativeEvent.keyCode === KeyCode.KEY_TAB) {
+      handleUpArrow();
+    } else if (event.nativeEvent.keyCode === KeyCode.KEY_TAB) {
+      handleDownArrow();
     }
   };
 
-  const renderMenuItems = (currentMenuItem, firstLevel) => {
+  const renderMenuItems = (currentMenuItem, parent = '', firstLevel = false) => {
     if (!currentMenuItem) {
       return undefined;
     }
 
     return currentMenuItem.map((item) => {
       const itemIsOpen = openKeys[item.path];
       const itemHasChildren = item.childItems !== undefined;
+      const id = item.name.split(' ').join('-');
       let isSelected = false;
       let selectedRef;
 
       if (selectedPath === item.path) {
         isSelected = true;
         selectedRef = selectedItem;
+      } else if (currentNodeId === id) {
+        isSelected = true;
       }
 
+      visibleNodes.current.push({ id, parent });
```

Comment:
When are these cleaned up? Does this change when items are scrolled out of view or is this for menu items that are rendered?

Reply:
visibleNodes is reinitialized to an empty array on each render. I just realized that there's no point to it being a ref in this case, so I made it a normal JS array.